python/cudf/cudf/core/indexed_frame.py

```diff
@@ -1695,6 +1696,191 @@ def last(self, offset):
             slice_func=lambda i: self.iloc[i:],
         )
 
+    @annotate("SAMPLE", color="orange", domain="cudf_python")
+    def sample(
+        self,
+        n=None,
+        frac=None,
+        replace=False,
+        weights=None,
+        random_state=None,
+        axis=None,
+        ignore_index=False,
+    ):
+        """Return a random sample of items from an axis of object.
+
+        You can use random_state for reproducibility.
+
+        Notes
+        -----
+        When sampling from ``axis=0/'index'``, ``random_state`` can be either
+        a numpy random state (``numpy.random.RandomState``) or a cupy random
+        state (``cupy.random.RandomState``). If a numpy random state is used,
+        the rows sampled is guarenteed to match pandas result, but performance
+        may be decreased if the item count is too high. Optionally, user may
+        specify a cupy random state to achieve better performance at high
+        item counts.
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of items from axis to return. Cannot be used with frac.
+            Default = 1 if frac = None.
+        frac : float, optional
+            Fraction of axis items to return. Cannot be used with n.
+        replace : bool, default False
+            Allow or disallow sampling of the same row more than once.
+            replace == True is not yet supported for axis = 1/"columns"
+        weights : numpy or cupy ndarray-like, optional
```

Comment:
I think we can safely just use ndarray. Our APIs should transparently handle the numpy<-->cupy casts where needed (or be able to rely on the underlying libraries).
```suggestion
        weights : ndarray-like, optional
```

Reply:
Not so simple. Cupy doesn't allow implicit conversion to numpy arrays so we might need to add more checks in current APIs if we allows cupy arrays. Numpy random states also cannot accept a cupy array so we need to assert type combinations of `weights` and `random_states`.

Reply:
Applied what we discussed offline today. We don't explicitly convert `weights` parameter but instead passes it into the `choice` method. Test fixtures are also updated to generate matching types of `weights` arrays.

Reply:
Sounds good.